execution/process.go

```diff
@@ -54,16 +59,45 @@ func (p *ProcessByKey) AddRecord(ctx context.Context, tx storage.StateTransactio
 	}
 
 	keyTuple := octosql.MakeTuple(key)
-	err = p.processFunction.AddRecord(ctx, tx, inputIndex, keyTuple, record)
-	if err != nil {
-		return errors.Wrap(err, "couldn't add record to process function")
-	}
 
 	eventTime := MaxWatermark
+	eventTimePrefixedTx := tx
 	if len(p.eventTimeField) > 0 {
 		eventTime = record.EventTime().AsTime()
+
+		// Adding record event time to keyTuple so that during Trigger prefixing process function is possible
+		keyTuple = octosql.MakeTuple(append([]octosql.Value{octosql.MakeTime(eventTime)}, key...))
+
+		// Adding keyTuple to event time storage
+		// For every record seen we store keys that appeared in that event time. It gives us an opportunity to
+		// clear/mark as fired old records, because both processFunction and trigger receives key or keyTuple to perform actions.
+		eventTimeMap := storage.NewMap(tx.WithPrefix(eventTimesSeenPrefix))
+
+		octoEventTime := octosql.MakeTime(eventTime)
+		var keysTuple octosql.Value
+		err := eventTimeMap.Get(&octoEventTime, &keysTuple)
+		if err == storage.ErrNotFound {
+			keysTuple = octosql.ZeroTuple()
+		} else if err != nil {
+			return errors.Wrap(err, "couldn't get records event time tuple from storage")
+		}
+
+		keysTuple = octosql.MakeTuple(append(keysTuple.AsSlice(), keyTuple.AsSlice()...))
```

Comment:
Secondly, this will completely decimate performance. The slice will grow for each key encountered, and each time you will load and set all keys. Please create a Set (we have those) prefixed by the event time.